### DISCUSSION TITLE
Will there be any documentation for how to interact w the COM Server published?

### DISCUSSION THREAD

**Discussion by byjrack** [microsoft/winget-cli#3953] [Q&A] [answered]
Bit of a long shot, but trying to figure out if I can streamline some scripting and since posh5 support seems to be up in the air interacting with the COM interface seems like the next best option.

Wasn't sure if it was just using https://learn.microsoft.com/en-us/uwp/api/windows.management.deployment.packagemanager or if there was more to the story.

**Comment by denelon** [comment, marked as answer]
There are a couple of specs you can use, and we will be publishing documentation. We've got a bit more work we want to do before updating the docs for these at Microsoft Learn.

* https://github.com/microsoft/winget-cli/blob/master/doc/specs/%23888%20-%20Com%20Api.md
* https://github.com/microsoft/winget-cli/blob/master/doc/specs/Configuration-COM-API.md

**Reply by denelon** [reply]
You can also look at the [SampleWinGetUWP caller](https://github.com/microsoft/winget-cli/tree/master/tools/SampleWinGetUWPCaller):

**Reply by byjrack** [reply]
yup cool.  Been out of the CPP game for a while so the wingetuwp i was banging my head against, but good to know i was generally in the right ballpark.  Figured if I can do some of the basics in CS than I can probably port any of that to another language with reasonable COM support.

**Reply by florelis** [reply]
Dev Home uses the winget COM API to install applications and is written in C#, maybe you could use it as a reference. You could start looking around [DevHome.SetupFlow.Common/WindowsPackageManager](https://github.com/microsoft/devhome/tree/630a274fc4628f64052e19306da515d528060ffc/tools/SetupFlow/DevHome.SetupFlow.Common/Configuration)

**Reply by byjrack** [reply]
Super helpful as have not needed to tie into COM for a very long time.  Gives me a fun project to do over downtime at the end of the year.

**Comment by vysakhs007** [comment]
Hey @denelon 
Do we support winget list from COM API? Or is there any way that we can use the currently exposed COM Interfaces to get list of applications installed in device?

**Reply by denelon** [reply]
Copy-paste from an internal response:

>packageManager.GetLocalPackageCatalog(LocalPackageCatalog::InstalledPackages) then FindPackages()?
Possibly with CreateCompositePackageCatalog to cross-reference remote sources

**Comment by vysakhs007** [comment]
Hey @denelon 
Does Winget COM API invocation is supported from a process running logged on as System Account?
I was able to invoke the COM APIs using Windows Admin Local user, but failed with System Account, Is this expected and are we going to have this support on future?

**Reply by denelon** [reply]
The Microsoft.WinGet.Client PowerShell modules work in system context. There are some challenges associated with doing that. I'll see if someone from the engineering team can add a bit of clarity.

**Reply by byjrack** [reply]
I know one of the reasons for me interacting w the COM interface as system or as user was because the cmdlets are still posh7+ and since the in-box version is 5 using them for orchestration hasn't worked.  I have had the same issues on @vysakhs007 in going DIY via the COM API, but given the cmdlets do it there is a path there just not a well worn one.  

I mean I would take the ability for a SYSTEM process to create a thread as the logged in user and tag it UAC'd, but I get that is a minefield for abuse.

**Reply by JohnMcPMS** [reply]
You cannot use the OOP COM server that ships in the MSIX package from SYSTEM.  This simply isn't supported on Windows currently.

You can use in-proc COM though: https://www.nuget.org/packages/Microsoft.WindowsPackageManager.InProcCom

**Reply by vysakhs007** [reply]
Hi all,
I would like to give more context on the issue that I am facing

- I have a Windows Service Process logged on as a Local System, from which I need to invoke winget COM API to fetch List of Applications installed in windows.
- Current approach that I used was using the following packages in my .NET Framework project
          - **Microsoft.WindowsPackageManager.ComInterop** , which contains the 
               -- _Microsoft.Management.Deployment.winmd_ reference to access the API and 
               -- `wintact.dll` used for activating winget server and retrieving COM instance for PackageManager and FindPackageOptions.
- The above approach works fine and I am able to retrieve list of application, but only when that process is logged in as a Local User Account.
- For System Account **WinGetServerManualActivation_CreateInstance** is throwing `HRESULT: 0x80070002 - File not found exception` , since the below unmanaged function is trying to access the _WindowsPackageMangager.exe_ located in my Local User App data folder.

```c++
HRESULT LaunchWinGetServerWithManualActivation()
 {
    const std::filesystem::path& localAppDataPath = GetKnownFolderPath(FOLDERID_LocalAppData);
    const std::filesystem::path& serverExePath = localAppDataPath / s_ServerExePath;
    std::wstring commandLineInput = std::wstring{ serverExePath } + L" --manualActivation";

    STARTUPINFO info = { sizeof(info) };
    wil::unique_process_information process;

    RETURN_LAST_ERROR_IF(!CreateProcessW(NULL, &commandLineInput[0], NULL, NULL, FALSE, 0, NULL, NULL, &info, &process));

    // Wait for manual reset event from server before proceeding with COM activation.
    wil::unique_event manualResetEvent;
    if (manualResetEvent.try_open(L"WinGetServerStartEvent"))
    {
        manualResetEvent.wait();
    }

    return S_OK;
}
````
So it would be really helpful if following questions can be answered

1. Is it possible to activate and access the COM API using the above method, if I start a new thread as user. @byjrack  Could you throw some more lights onto this? I hope only the ManualServer Activation requires the User Context or Is the COM Object creation also requires User Context?
2. @JohnMcPMS The InProc COM that you are talking about - Is it possible to use for my scenario and could you provide some documentation on how we can use it? _https://github.com/microsoft/winget-cli/blob/master/doc/specs/%23888%20-%20Com%20Api.md_  - This doc says that it should be access as Out-OfProc COM, so I am little bit confused here.

Note: I don't want the Powershell/CLI to be invoke from my process, the only way is to invoke COM API only from a System Process.

**Reply by JohnMcPMS** [reply]
The COM server requires the user to be "logged in" to be activated, which happens (enough) through the service running as a user to count.  Simply impersonating the user on a thread without any other method is not likely to work, even for inproc COM, since it relies on reading the user's registry hive and that won't be mounted.

We support running in a user session and running as `SYSTEM` in any session (via PowerShell / inproc COM).  Running as the user in session 0 (aka as a service) is not supported, although you are not the first to ask for it this week.  One can certainly make it work to some degree, but we do not have any test infrastructure to ensure that it does work.

If you are ok with enumerating the installed packages as `SYSTEM`, then the inproc COM will work.  We are woefully underdocumented there unfortunately, but I do have plans to improve that in the near future.

**Reply by vysakhs007** [reply]
@JohnMcPMS Thanks for the reply
So you are suggesting inproc COM should work  in a **SYSTEM** context to access the COM Objects and enumerate the list of installedApplications (similar to that of in Add or Remove Programs)? 
Also I would like to know is there any other way to get a list of installed applications (Similar to the ONE listed in Add or Remove Programs) in windows?

**Reply by Niranjan-Jyothi** [reply]
Hey @JohnMcPMS any plan for the documentation of inproc COM in c#? I'm facing issues initialising objects (PackageManager etc) and consuming APIs; would be really helpful if you could point to a simple code example for the time being. I did have a look at the Powershell reference within this repo but could not find any class/interface registration to be done before object initialising)
cc: @denelon @Trenly

**Reply by PatrickSchmidtSE** [reply]
@vysakhs007 
did you find a solution ? we try to archive the exact same thing. Via the CLI it can be solved, with a workaround by calling the RIGHT winget.exe . Installed Apps machinewide will be displayed for the LOCAL SYSTEM account (service). But with the COM-API we are stuck with the same problems you have...

**Comment by denelon** [comment]
We will be providing a sample for C# and updating the documentation at Microsoft Learn. We're getting a pipeline set up to export the API for Microsoft Learn.

* https://github.com/microsoft/winget-cli/issues/4094
